TM-4910-Bureau-Exception-List-API-Integration

cagbakwu opened the pull request:
Dual Merge: https://github.com/MetaPhase-Consulting/State-TalentMAP-API/pull/852

https://metaphase.atlassian.net/browse/TM-4910
Integrated APIs into Bureau Exception List

cagbakwu commented on src/actions/bureauException.js:
```diff
@@ -280,3 +188,27 @@ export function saveBureauExceptionSelections(data) {
   };
 }
+
+export function deleteBureauExceptionList(data) {
```
Wondering, could I leave these as post since we arent actually doing the deleting and just pushing the data to web services.

cagbakwu commented on src/actions/bureauException.js:
```diff
@@ -257,11 +141,35 @@ export function bureauExceptionPositionEditSuccess(data) {
+  };
+}
+
 export function saveBureauExceptionSelections(data) {
```
Wondering, could I leave these as post since we aren't actually doing the updating and just pushing the data to web services.

elizabeth-jimenez commented on src/Components/AdministratorPage/BureauExceptionList/BureauExceptionListCard.jsx:
```diff
@@ -156,50 +181,67 @@ const BureauExceptionListCard = (props) => {
-  id: PropTypes.number.isRequired,
-  Name: PropTypes.string.isRequired,
+  userData: PropTypes.shape({
+    bureauCode: PropTypes.string,
+    bureaus: PropTypes.string,
```
I thought the plan was to make these arrays at the api layer?

jlam-mpc commented:
Left a note on your API branch already but we deployed it to dev1 and it's receiving 500 responses from your service code instead of FSBID so it would be good to revisit the API code and make sure it's not returning any errors other than not being able to reach the FSBID EP. That's the only thing we should be seeing if you don't have a mock PR I believe.

jlam-mpc commented:
![image](https://github.com/MetaPhase-Consulting/State-TalentMAP/assets/121953273/904d48c4-98ea-4f88-8fc2-7f1c0b405a43)

As you can see, it's returning the message "string indices must be integers".

M00staff commented on src/actions/bureauException.js:
```diff
@@ -2,143 +2,21 @@ import { batch } from 'react-redux';
-
 export function bureauExceptionFetchDataErrored(bool) {
   return {
     type: 'BUREAU_EXCEPTION_FETCH_HAS_ERRORED',
```
this action doesn't do anything

M00staff replied:
![image](https://github.com/MetaPhase-Consulting/State-TalentMAP/assets/12723877/19cbd398-c137-4ff2-9e59-64fe88ebabae)
![image](https://github.com/MetaPhase-Consulting/State-TalentMAP/assets/12723877/60371732-c83e-4c0a-992e-196a7d2bcbda)

M00staff commented on src/actions/bureauException.js:
```diff
@@ -202,35 +90,31 @@ export function bureauExceptionFetchData(query = {}) {
 
-export function bureauExceptionBureauDataFetchData(query = {}) {
+export function bureauExceptionBureauDataFetchData(currentUser) {
+  console.log('currentUser', currentUser);
```
```suggestion
```

M00staff commented on src/actions/bureauException.js:
```diff
@@ -148,7 +26,21 @@ export function bureauExceptionFetchDataErrored(bool) {
+
+export function bureauExceptionBureauDataFetchDataLoading(bool) {
+  return {
+    type: 'BUREAU_EXCEPTIONS_BUREAU_DATA_FETCH_IS_LOADING',
```
you're mixing singular & plural quite a bit here - the `type` actually needs to match what is in the reducer, other wise it will not call the correct function... 

Meaning - this does not work and isn't wired up correctly.

M00staff commented:
this crashes both locally and on dev - its not finished...

M00staff commented:
got this running locally (make sure to link the mock PR to the other PRs) and noticed some strange behavior

when both of these tabs are open and you click on "Bureau" - it activates the other tab's bureau checkbox... i think there is some kind of mix up with the checkbox's state

<img width="304" alt="image" src="https://github.com/MetaPhase-Consulting/State-TalentMAP/assets/12723877/abe356cf-4fbf-489c-88c4-c279db9c3e24">

elizabeth-jimenez commented:
![image](https://github.com/MetaPhase-Consulting/State-TalentMAP/assets/23065001/bbb61ea1-e544-4e31-b783-dfc5eaafbc5c)
let's match the names up

elizabeth-jimenez commented:
![image](https://github.com/MetaPhase-Consulting/State-TalentMAP/assets/23065001/1e5f3132-214a-49c2-9d00-1a19df719539)
when we talked about this, I thought we had agreed to hide this from the user and just present them with a Save or Cancel

elizabeth-jimenez commented:
![image](https://github.com/MetaPhase-Consulting/State-TalentMAP/assets/23065001/85833819-1c52-4849-b003-249d1dbb9ad1)
opening up the Edit adds a lot of blank space to the end of page

elizabeth-jimenez commented on src/Constants/SystemMessages.js:
```diff
@@ -263,6 +263,11 @@ export const BUREAU_EXCEPTION_POSITION_EDIT_SUCCESS = 'The Bureau Exception Posi
+export const DELETE_BUREAU_EXCEPTION_SUCCESS_TITLE = 'Bureau Exception Deleted';
+export const DELETE_BUREAU_EXCEPTION_SUCCESS = 'The Bureau Exception has been deleted';
+export const DELETE_BUREAU_EXCEPTION_ERROR_TITLE = 'Bureau Exception Delete Error';
+export const DELETE_BUREAU_EXCEPTION_ERROR = 'There was an issue attempting to delete the selected Bureau Exception. Please try again.';
```
By the time this PR is done, we may only need an update success and save. Make sure to keep the established pattern of :
Success TITLE
Success Message
Error TITLE
Error message

elizabeth-jimenez commented on src/Constants/SystemMessages.js:
```diff
@@ -263,6 +263,11 @@ export const BUREAU_EXCEPTION_POSITION_EDIT_SUCCESS = 'The Bureau Exception Posi
 export const BUREAU_EXCEPTION_POSITION_EDIT_ERROR_TITLE = 'Bureau Exception Position Edit Error';
 export const BUREAU_EXCEPTION_POSITION_EDIT_ERROR = 'There was an issue attempting to update this Bureau Exception Position. Please try again.';
```
`BUREAU_EXCEPTION_POSITION_EDIT_ERROR` the `POSITION` in these, is that accurate? Wondering if the deviation from the other Bureau Exception messages is intentional

elizabeth-jimenez commented:
`bureauExceptionsBureauDataFetchDataErrored` Try to do your best at naming - it may not seem like a big deal when setting it up, but when there is a time crunch and we are debugging, small things like good naming practices help a ton

elizabeth-jimenez commented:
Piggy backing on what Moose said - there seems to be a lot of redux that is not being used.
verify if you need those states at all and:
- if not needed, remove from everywhere
- or keep and make sure it is named well and properly hooked up

cagbakwu commented:
> ![image](https://user-images.githubusercontent.com/23065001/278075533-bbb61ea1-e544-4e31-b783-dfc5eaafbc5c.png) let's match the names up

@elizabeth-jimenez 
Just confirming, the AC for this was labeled this way. What are your thoughts?

https://metaphase.atlassian.net/browse/TM-4834

elizabeth-jimenez commented:
> > ![image](https://user-images.githubusercontent.com/23065001/278075533-bbb61ea1-e544-4e31-b783-dfc5eaafbc5c.png) let's match the names up
> 
> @elizabeth-jimenez Just confirming, the AC for this was labeled this way. What are your thoughts?
> 
> https://metaphase.atlassian.net/browse/TM-4834

i'm talking about the discrepancy between the **menu item and icon** and the **page title and icon**

elizabeth-jimenez commented:
For your Action file, check out the new pattern we are implementing in `src/actions/jobCategories.js` and `src/actions/publishablePositions.js`. Notice how we just have redux states for our reads and for our action calls(create, delete, edit, etc), we are firing off Toasts to communicate success or failure to our users and then re-routing by calling the relevant Read actions -> which in turn have their own load, error, and success redux states that we can use

cagbakwu commented:
- [ ] Updated naming convention in Redux
- [ ] updated redux calls for each action
- [ ] Updated Props for Card
- [ ] Updated Checkboxes to select properly
- [ ] Updated header name to match sidebar
- [ ] Removed extra space at the bottom of final card.
- [ ] Updated names of constants

elizabeth-jimenez commented:
> @elizabeth-jimenez
> 
> * [x]  Updated naming convention in Redux
> * [x]  updated redux calls for each action
> * [x]  Updated Props for Card
> * [x]  Updated Checkboxes to select properly
> * [x]  Updated header name to match sidebar
> * [x]  Removed extra space at the bottom of final card.
> * [x]  Updated names of constants

i'm still not seeing some of the feedback addressed

elizabeth-jimenez commented:
Before you open this back up for review, make a more extensive list of the feedback, to really make sure you've addressed it all before someone spends time coming back to review this. Copy/paste the requested changes and for each one explain how you've addressed it.

elizabeth-jimenez commented:
Another thing you'll want to address is the cancel token - i sent a message in tm-devs about it, go check it out to see how to update

cagbakwu commented:
- [x] Updated naming convention of actions and reducers in Redux. 'actions/bureauExceptions' 'reducers/bureauExceptions'
- [x] updated redux calls for 'bureauExceptionListCard' and linked the appropriate action.
- [x] Updated 'bureauCodes' and 'bureaus' Prop types from strings to arrays.
- [ ] removed console.logs in 'bureauExceptionListCard'.
- [ ] consolidated the 'Save', 'Update', and 'Delete' buttons in card views into one button.
- [x] Updated Checkboxes to correctly selection all and none in 'bureauExceptionListCard'. 
- [x] Updated header name to match sidebar link.
- [x] Removed extra height space below card display list.
- [x] Updated names of constants in system messages for 'Add', 'Update', and 'Delete' toasts for  'actions/bureauExceptions'.
- [ ] Updated how cancel token is implemented in api calls in  'actions/bureauExceptions'.

elizabeth-jimenez commented on src/Constants/SystemMessages.js:
```diff
@@ -248,20 +248,20 @@ export const BIDDER_PORTFOLIO_ADD_SUCCESS = 'Bidder Portfolio has been added.';
-export const BUREAU_EXCEPTION_POSITION_EDIT_SUCCESS = 'The Bureau Exception Position has been updated';
-export const BUREAU_EXCEPTION_POSITION_EDIT_ERROR_TITLE = 'Bureau Exception Position Edit Error';
-export const BUREAU_EXCEPTION_POSITION_EDIT_ERROR = 'There was an issue attempting to update this Bureau Exception Position. Please try again.';
+export const BUREAU_EXCEPTION_ADD_ERROR = 'There was an issue attempting to add this Bureau Exception. Please try again.';
```
```suggestion
export const BUREAU_EXCEPTION_ADD_ERROR = 'There was an issue attempting to add Bureau Exceptions. Please try again.';
```

elizabeth-jimenez commented on src/Constants/SystemMessages.js:
```diff
@@ -248,20 +248,20 @@ export const BIDDER_PORTFOLIO_ADD_SUCCESS = 'Bidder Portfolio has been added.';
-export const BUREAU_EXCEPTION_POSITION_EDIT_ERROR_TITLE = 'Bureau Exception Position Edit Error';
-export const BUREAU_EXCEPTION_POSITION_EDIT_ERROR = 'There was an issue attempting to update this Bureau Exception Position. Please try again.';
+export const BUREAU_EXCEPTION_ADD_ERROR = 'There was an issue attempting to add this Bureau Exception. Please try again.';
+export const BUREAU_EXCEPTION_ADD_ERROR_TITLE = 'Bureau Exception Add Error';
+export const BUREAU_EXCEPTION_ADD_SUCCESS = 'Bureau Exception has been added.';
+export const BUREAU_EXCEPTION_ADD_SUCCESS_TITLE = 'Bureau Exception Added';
+
+export const BUREAU_EXCEPTION_EDIT_SUCCESS_TITLE = 'Bureau Exception Updated';
+export const BUREAU_EXCEPTION_EDIT_SUCCESS = 'The Bureau Exception has been updated';
+export const BUREAU_EXCEPTION_EDIT_ERROR_TITLE = 'Bureau Exception Edit Error';
+export const BUREAU_EXCEPTION_EDIT_ERROR = 'There was an issue attempting to update this Bureau Exception. Please try again.';
+
+export const BUREAU_EXCEPTION_DELETE_SUCCESS_TITLE = 'Bureau Exception Deleted';
+export const BUREAU_EXCEPTION_DELETE_SUCCESS = 'The Bureau Exception has been deleted';
+export const BUREAU_EXCEPTION_DELETE_ERROR_TITLE = 'Bureau Exception Delete Error';
+export const BUREAU_EXCEPTION_DELETE_ERROR = 'There was an issue attempting to delete the selected Bureau Exception. Please try again.';
```
let's update all of these messages to be plural for Bureau Exceptions - variable names are good

elizabeth-jimenez replied:
keeping Bureau Exception for the Titles are fine, matches our page title

elizabeth-jimenez commented:
Im still seeing weird checkbox issues

elizabeth-jimenez commented:
There is also sometimes a jumping action - jumps to bottom of page - when selecting a checkbox

elizabeth-jimenez commented on src/Components/AdministratorPage/BureauExceptionList/BureauExceptionListCard.jsx:
```diff
@@ -1,30 +1,54 @@
+  const [bureauCodes, setBureauCodes] = useState([]);
+
+  const gatherBureauCodes = () => {
+    if (bureauCodeList !== null && bureauCodeList !== undefined && bureauCodeList !== ' ') {
```
what are you trying to handle with this line?

cagbakwu replied:
@elizabeth-jimenez  This code handles if you select options in the checkboxes and then decide to cancel. It puts back the originally check bureaus. I renamed it to gatherInitialBureauCodes.

elizabeth-jimenez commented on src/Components/AdministratorPage/BureauExceptionList/BureauExceptionListCard.jsx:
```diff
@@ -34,17 +58,33 @@ const BureauExceptionListCard = (props) => {
-    e.preventDefault();
+  const addBureaus = () => {
     const currentUser = {
+      bureauCode: bureauCodes,
       id,
-      Name,
-      bureauCode,
+    };
+    dispatch(addBureauExceptionSelections(currentUser));
+  };
+
+  const deleteBureaus = () => {
+    const currentUser = {
+      id,
+      pv_id,
+    };
+    dispatch(deleteBureauExceptionList(currentUser));
+  };
+
+  const modify = () => {
+    const currentUser = {
+      bureauCode: bureauCodes,
+      id,
+      pv_id,
     };
```
i may be seeing a discrepancy here between our FE and BE - on the BE, we are passing through a lot more data than code, id, and pv_id to the WS procedure - but we don't seem to be passing it along here

elizabeth-jimenez commented:
the naming of Bureau Exception List vs Bureau Exception isn't very informative. Let's rename to something more like your API naming `bureau_exception_list_bureaus` for the one where you pull the Bureau ref data

cagbakwu commented:
> There is also sometimes a jumping action - jumps to bottom of page - when selecting a checkbox
@elizabeth-jimenez 
The comments referring to the checkboxes happened because the ids needed to be changed. Its resolved now.

cagbakwu commented on src/reducers/bureauException/bureauException.js:
```diff
@@ -1,31 +1,49 @@
-export function bureauExceptionsOptions(state = [], action) {
+export function bureauExceptionListErrored(state = false, action) {
+  switch (action.type) {
+    case 'BUREAU_EXCEPTION_LIST_IS_ERRORED':
```
Changed

cagbakwu commented:
- [x] Updated naming convention of actions and reducers in Redux. 'actions/bureauExceptions' 'reducers/bureauExceptions'
- [x] updated redux calls for 'bureauExceptionListCard' and linked the appropriate action.
- [x] removed excess redux calls that were not used.
- [x] Updated 'bureauCodes' and 'bureaus' Prop types from strings to arrays.
- [x] removed console.logs in 'bureauExceptionListCard'.
- [x] consolidated the 'Save', 'Update', and 'Delete' buttons in card views into one button.
- [x] Updated Checkboxes to correctly selection all and none in 'bureauExceptionListCard'. 
- [x] Updated header name to match sidebar link.
- [x] Removed extra height space below card display list.
- [x] Updated names of constants in system messages for 'Add', 'Update', and 'Delete' toasts for  'actions/bureauExceptions'.
- [x] Updated how cancel token is implemented in api calls in  'actions/bureauExceptions'.

New
- [x] Renamed actions for retrieving users list and user bureaus list
- [ ] changed 'BUREAU_EXCEPTION_LIST_IS_ERRORED' to 'BUREAU_EXCEPTION_LIST_HAS_ERRORED' 
- [ ] Adjust checkbox id's to resolve multiples being selected and screen jumps.

elizabeth-jimenez commented:
![Pasted Graphic 7](https://github.com/MetaPhase-Consulting/State-TalentMAP/assets/23065001/dafb0534-813e-4d6b-bf8a-0a2d17fc891b)

Check out how it differs from dev branch UI with the side margins for cards
![Pasted Graphic 6](https://github.com/MetaPhase-Consulting/State-TalentMAP/assets/23065001/5f7426bb-715f-4cc3-9fbf-ec0683378ccd)

elizabeth-jimenez commented:
![image](https://github.com/MetaPhase-Consulting/State-TalentMAP/assets/23065001/396d28a3-0f64-458e-bbf2-23b7af6c2e77)
Spinner is huge - i see some stylings for 100px set, but i dont think they are getting grabbed

elizabeth-jimenez commented on src/Components/AdministratorPage/BureauExceptionList/BureauExceptionList.jsx:
```diff
@@ -4,28 +4,27 @@ import { useDispatch, useSelector } from 'react-redux';
+  const noResults = BureauExceptionData?.length === 0;
   const getOverlay = () => {
     let overlay;
     if (BureauExceptionDataLoading) {
```
for spinner, just replace "bid-season-filters" with "standard-center"

elizabeth-jimenez commented on src/sass/_bureauExceptionList.scss:
```diff
@@ -44,9 +44,8 @@
-  min-height: 100vh;
-  overflow-y: scroll;
+.bel-lower-section {
+  overflow-y: clip;
```
```suggestion
  overflow-y: clip;
  margin-inline: 20px;
```

cagbakwu replied:
Adding styling properties from similar cards  to match.

elizabeth-jimenez commented:
> ![Pasted Graphic 7](https://user-images.githubusercontent.com/23065001/280379647-dafb0534-813e-4d6b-bf8a-0a2d17fc891b.png)
> 
> Check out how it differs from dev branch UI with the side margins for cards ![Pasted Graphic 6](https://user-images.githubusercontent.com/23065001/280379740-5f7426bb-715f-4cc3-9fbf-ec0683378ccd.png)

left a code suggestion for this

elizabeth-jimenez commented:
stil seeing the jumping action on some checkboxes - checkout CSO one

elizabeth-jimenez commented on src/Components/AdministratorPage/BureauExceptionList/BureauExceptionListCard.jsx:
```diff
@@ -132,7 +188,7 @@ const BureauExceptionListCard = (props) => {
                       <TextInput
                         changeText={(e) => setBureau(e)}
-                        label={`F/S Employee Name: ${Name}`}
+                        label={`F/S Employee Name: ${name}`}
```
let's delete this label: the name is already shown right above and it doesnt make sense to have the name as a label for the input of filtering bureaus

elizabeth-jimenez commented:
![image](https://github.com/MetaPhase-Consulting/State-TalentMAP/assets/23065001/90f054b6-a9a8-45f0-a3ac-d547fcabea90)
card still getting cut off before 1100px

elizabeth-jimenez commented on src/Components/AdministratorPage/BureauExceptionList/BureauExceptionList.jsx:
```diff
@@ -4,32 +4,30 @@ import { useDispatch, useSelector } from 'react-redux';
-  const BureauExceptionData = useSelector(state => state.bureauExceptions);
-  const BureauExceptionOptionsData = useSelector(state => state.bureauExceptionsOptions);
-  const BureauExceptionError = useSelector(state => state.bureauExceptionsFetchDataErrored);
+  const BureauExceptionDataLoading = useSelector(state => state.bureauExceptionLoading);
+  const BureauExceptionData = useSelector(state => state.bureauExceptionSuccess);
+  const BureauExceptionError = useSelector(state => state.bureauExceptionErrored);
```
🙌

elizabeth-jimenez commented on src/Components/AdministratorPage/BureauExceptionList/BureauExceptionList.jsx:
```diff
@@ -4,32 +4,30 @@ import { useDispatch, useSelector } from 'react-redux';
+  const BureauExceptionError = useSelector(state => state.bureauExceptionErrored);
   const fetchAndSet = () => {
-    dispatch(bureauExceptionFetchData());
-    dispatch(bureauExceptionBureauDataFetchData());
+    dispatch(bureauExceptionUsersListFetchData());
   };
```
fetchAndSet just being used as a pass-through. Move dispatch to useEffect

elizabeth-jimenez commented on src/Components/AdministratorPage/BureauExceptionList/BureauExceptionList.jsx:
```diff
@@ -43,20 +41,18 @@ const BureauExceptionList = () => {
       </div>
       {
         getOverlay() ||
         <>
```
don't need a fragment if you only have a single element

elizabeth-jimenez commented:
![image](https://github.com/MetaPhase-Consulting/State-TalentMAP/assets/23065001/869e5508-06d1-4792-86cb-47a1a7eea577)
Ask WS about this data node - we may want to filter it out based on what they say

cagbakwu commented:
- [ ] Removed fragments from overlay call
- [ ] Adjusted the css for the card with down to 1100px
- [ ] replaced currentUserInfo with the actual request loading state
- [ ] removed fetch and set and called dispatch from useEffect

elizabeth-jimenez commented:
![image](https://github.com/MetaPhase-Consulting/State-TalentMAP/assets/23065001/97834fd6-3dd0-45ad-a66b-9feb43c6149d)
nice

elizabeth-jimenez commented:
Future Fix: When i hit Edit I get a load, i then select a few bureaus, then i hit close, then hit Edit again and get another load, but those previously checked (but not saved) checkboxes are still checked - the load should mean it is re-grabbing the users Bureaus, but it seems to be keeping the local changes 🤷‍♀️

cagbakwu commented:
> Future Fix: When i hit Edit I get a load, i then select a few bureaus, then i hit close, then hit Edit again and get another load, but those previously checked (but not saved) checkboxes are still checked - the load should mean it is re-grabbing the users ~Bureaus~ data, but it seems to be keeping the local changes 🤷‍♀️
> 
> On more thought, more a UX thing, since the load is for Data the user isn't even presented with

@elizabeth-jimenez  I just added the call to reset the checkboxes

elizabeth-jimenez commented on src/Components/AdministratorPage/BureauExceptionList/BureauExceptionList.jsx:
```diff
@@ -4,32 +4,26 @@ import { useDispatch, useSelector } from 'react-redux';
+      overlay = <Spinner type="standard-center" class="homepage-position-results" size="medium" />;
     } else if (BureauExceptionError) {
       overlay = <Alert type="error" title="Error loading results" messages={[{ body: 'Please try again.' }]} />;
     } else if (noResults) {
```
the info overlay message needs to be updated "Please broaden your search criteria and try again." not relevant to page bc no filters
